src/watershed.jl

```diff
@@ -0,0 +1,97 @@
+import Base.isless
+
+immutable PixelKey{CT}
+    val::CT
+    time_step::Int
+end
+isless{T}(a::PixelKey{T}, b::PixelKey{T}) = (a.val < b.val) || (a.val == b.val && a.time_step < b.time_step)
+
+"""
+```
+segments                = watershed(img, [markers])
+``` 
+Segments the image using watershed transform. Each basin formed by watershed transform corresponds to a segment.
+If no markers are provided, local minima of the image are taken as markers.
+    
+Parameters:  
+-    img            = input grayscale image
+-    markers        = an array marking the pixels where flooding should start
+
+"""
+function watershed{T<:Images.NumberLike}(img::Array{T, 2}, markers::Array{T,2})
+
+    segments = copy(markers)
+    pq = PriorityQueue(CartesianIndex{2}, PixelKey{T}, Base.Order.Forward)
+    time_step = 0
+
+    R = CartesianRange(size(img))
+    Istart, Iend = first(R), last(R)
+    for i in CartesianRange(size(img))
+        if markers[i] > 0
+            for j in CartesianRange(max(Istart,i-1), min(i+1,Iend))
+                if segments[j] == 0
+                    segments[j] = markers[i]
+                    enqueue!(pq, j, PixelKey(img[i], time_step))
+                    time_step += 1
+                end
+            end
+        end
+    end
+
+    while length(pq) > 0
+        current = dequeue!(pq)
+        for j in CartesianRange(max(Istart,current-1), min(current+1,Iend))
+            if segments[j] == 0
+                segments[j] = segments[current]
```

Comment:
hoist this array access? (would be nice if the compiler did that for you, but I don't think it does ATM)